paasta_tools/remote_git.py

```diff
@@ -23,13 +23,29 @@ def _make_determine_wants_func(ref_mutator):
     determine_wants argument to dulwich's send_pack method. The returned
     function will not delete or modify any existing refs."""
     def determine_wants(old_refs):
-        refs = dict(old_refs)
+        refs = {k.decode('UTF-8'): v.decode('UTF-8') for k, v in old_refs.items()}
         new_refs = ref_mutator(refs)
+        new_refs = {k.encode('UTF-8'): v.encode('UTF-8') for k, v in new_refs.items()}
         new_refs.update(old_refs)  # Make sure we don't delete/modify anything.
         return new_refs
     return determine_wants
 
 
+def make_force_push_mutate_refs_func(targets, sha):
+    """Create a 'force push' function that will inform send_pack that we want
+    to mark a certain list of target branches/tags to point to a particular
+    git_sha.
+
+    :param targets: List of branches/tags to point at the input sha
+    :param sha: The git sha to point the branches/tags at
+    :returns: A function to do the ref manipulation that a dulwich client can use"""
+    def mutate_refs(refs):
+        for target in targets:
+            refs[target.encode('UTF-8')] = sha.encode('UTF-8')
+        return refs
+    return mutate_refs
+
+
 def create_remote_refs(git_url, ref_mutator, force=False):
```

Comment:
So I think the real problem is that anyone calling this can pass their own `ref_mutator`. And it turns out this is exactly what `paasta stop` does.

Could we wrap another function to do the encoding? i.e. like `_make_determine_wants_func` but for every call to `create_remote_refs` not just when `force is False`. That way we could just have the encoding in one place and it would catch the `paasta stop` usage too.

Also, if it's easy and you have time can you add a similar test for `paasta stop`. I haven't looked in detail but it's definitely broken on master so we shouldn't have a green build. I started work on some higher level itests that may or may not be a good idea (#1039) but it would be nice to have an itest like the one you added here for mark-for-deployment.